### ISSUE TITLE
Galaxy Tab S9 Ultra Touch Pad Compatible Issues

### ISSUE THREAD

**Issue by hansm629** [termux/termux-x11#434]
With the `Capture external mouse when possible` option active

These are issues that occurred on the Galaxy Tab S9 Ultra Keyboard Cover.

As of `PRoot-Distro Ubuntu 23.04`.

**1. All gestures, including one-tab gestures, do not work**
Galaxy Tab S8+ Keyboard Cover Touchpad used gestures
but Galaxy Tab S9 Ultra Touchpad doesn't work with all gestures, including one tap.

**2. Too fast mouse pointer speed**
When you use the touchpad on the Galaxy Tab S9 Ultra Keyboard Cover, the mouse pointer is incredibly fast to an uncontrollable degree.
Adjusting mouse sensitivity in XFCE does not apply.

Please check.
[log_20230823_3.txt](https://github.com/termux/termux-x11/files/12419961/log_20230823_3.txt)

**Comment by SanwariyaAmit** [comment]
Have the same issue on my Tab S9+, to click I have to press the touchpad, which is horrible on the book cover keyboard and moves the cursor, so I end up clicking someplace else. The two-finger scroll doesn't work at all, neither does drag and drop.

**Comment by ilkersigirci** [comment]
Also have the same problem in my [termux_xfce](https://github.com/phoenixbyrd/Termux_XFCE) setup. I hope it can be fixed.

**Comment by greentail-kitsune** [comment]
Same problem. If the cursor direction and speed can be changed in the input matrix. Then two-finger gestures are not corrected in any way.
Any chance this will be fixed?

**Comment by mistletoe235** [comment]
I have the same problem. And I have spent two hours to change the source code. You can leave your email and I'll send it. The version was added with "dex". As for me, all the functions of a touchpad work all right.
![{115519-x-11080731504393}](https://github.com/termux/termux-x11/assets/92664458/169304a1-56e7-49d8-b282-cac852c7096b)

**Comment by twaik** [comment]
@mistletoe235 you can post it here as zip or tarball.

**Comment by mistletoe235** [comment]
I have the same problem. And I have spent two hours to change the source code. The version was added with "dex". As for me, all the functions of a touchpad work all right.

[apk_and_changed_code.zip](https://github.com/termux/termux-x11/files/14067311/apk_and_changed_code.zip)

**Comment by twaik** [comment]
Priobably it will be better to make a pull request. This way we will see all the changes and system will generate updated apk.

**Comment by Knightly-Spartan** [comment]
Would be easier for users to upgrade too.

**Comment by twaik** [comment]
Upgrade what?

**Comment by Knightly-Spartan** [comment]
I meant update the app. The apk from the zip gets recognized as different app and Android is refusing to do update the app.

**Comment by twaik** [comment]
I will merge the changes only after making sure it does not break existing code for other users. I am not refusing to upgrade it.

**Comment by mistletoe235** [comment]
Sorry, I forgot to sign it with the official key. If you encountered any problems due to the key, you can change to this apk file. @Knightly-Spartan 
[app-arm64-with-same-sign.zip](https://github.com/termux/termux-x11/files/14070448/app-arm64-with-same-sign.zip)

**Comment by twaik** [comment]
@mistletoe235 pull request please.

**Comment by greentail-kitsune** [comment]
@mistletoe235 Have you succeeded in resolving the two-finger gesture problem? I've tested your fix, but the swipe direction with two fingers is still malfunctioning.
The most detailed description of the problem is at the link
https://github.com/termux/termux-x11/issues/419

 Galaxy Tab S9 Ultra

**Comment by mistletoe235** [comment]
@greentail-kitsune 
I only debug on my S8+, maybe S9 Ultra has some else problems.
The problem that video shows I only find when I connect the mouse and using the touchpad.
I also don't know how to fix it, sorry.

**Comment by AlphaBs** [comment]
I found that I can drag with two fingers. Clicking on the left side of the touchpad with one finger (not touching) and moving it with the other finger is dragging.

**Comment by mjelintacharge** [comment]
@twaik 
Galaxy Tab S9 Ultra user here, can confirm this patch by @Knightly-Spartan does indeed fix annoying things about samsung book cover keyboard touch pad:

1. Tapping to perform a left click now works fine
2. Holding left click and then dragging something now works (finally!)

However!
This is all with capture external mouse when possible disabled.
Enabling this option does still result in touch pad being unusable.

**Comment by ilkersigirci** [comment]
It's been 7 months since the issue is created. Although related PR's exists, they seem to not going to merge with the main branch. Will there be any official solution to this problem?

**Comment by twaik** [comment]
Again. I am not going to merge PR because it breaks compqtibility with existing devices. 
And I can not fix this PR because I have no devices for tests.

**Comment by ilkersigirci** [comment]
> Again. I am not going to merge PR because it breaks compqtibility with existing devices. 
> And I can not fix this PR because I have no devices for tests.

I can help with testing if requested. It would be awesome if this is fixed for our devices

**Comment by twaik** [comment]
> I can help with testing if requested. 

Testing means testing in real time. I can not make changes, send apk and wait for response.

**Comment by ilkersigirci** [comment]
Alright then, just wanted to help

**Comment by AlphaBs** [comment]
maybe I can fix the issue. I have a device and I can test with it. but I'm not sure it's okay to copy the original PR to open new PR. this would lose the author of the original commit.

**Comment by ilkersigirci** [comment]
I have installed the latest build on my Tab S9 Ultra. Unfortunately, tap to click or touchpad two finger gestures not working. Hence, I don't think this issue should be closed

**Comment by twaik** [comment]
@ilkersigirci check it with "Capture external mouse" disabled...

**Comment by ilkersigirci** [comment]
I have checked and tried all the combinations in the settings. Unfortunately, none of them seems to be working, sir

**Comment by twaik** [comment]
@AlphaBs can you please help?

**Comment by AlphaBs** [comment]
@twaik looks like the commit was reverted? 76309c1c7b7a515a942a650d2b8e8e40fd06c865

@ilkersigirci can you try this build? https://github.com/termux/termux-x11/actions/runs/8410806031

**Comment by AlphaBs** [comment]
@twaik looks like the commit was reverted? 76309c1c7b7a515a942a650d2b8e8e40fd06c865

@ilkersigirci can you try this build? https://github.com/termux/termux-x11/actions/runs/8410806031

my setting is disabled "Capture External Mouse" and "Trackpad" mode.

**Comment by ilkersigirci** [comment]
I have tried it and it seems to working perfectly. I am using the same settings that you've mentioned

**Comment by AlphaBs** [comment]
About "Capture External Mouse" option, on my device just skipping [this code](https://github.com/termux/termux-x11/blob/76309c1c7b7a515a942a650d2b8e8e40fd06c865/app/src/main/java/com/termux/x11/input/TouchInputHandler.java#L185) resolves the issue. 
I can't debug the exact reason since I can't use my device for now, what about to add a new option to skip it?

**Comment by twaik** [comment]
> on my device just skipping [this code](https://github.com/termux/termux-x11/blob/76309c1c7b7a515a942a650d2b8e8e40fd06c865/app/src/main/java/com/termux/x11/input/TouchInputHandler.java#L185) resolves the issue.

This code is needed to handle dex touchpad (or any other touchpad that send touchscreen events) as a touchpad even in the case if user chooses "Simulated touchscreen" or "Direct touchscreen" modes. 

So skipping that code will brake dex touchpad in the case user chooses something other that "Trackpad" touch mode. That is not an option.

**Comment by Knightly-Spartan** [comment]
Maybe a toggle to skip the code or not. So it works for both DeX tounchpad and the keyboard trackpads

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by twaik** [comment]
> Maybe a toggle to skip the code or not.

Again. Skipping that code is not an option. It will break dex touchpad in the case if option other than "Trackpad" will be chosen. We should fix that code, not skip it.

**Comment by mjelintacharge** [comment]
@twaik 

Why is this issue closed, if the commit regarding this issue (https://github.com/termux/termux-x11/commit/2104ea690d3f043d381f2ca9c6e6cfef723b4ca5) was reverted back via this commit (https://github.com/termux/termux-x11/commit/76309c1c7b7a515a942a650d2b8e8e40fd06c865)??

What needs to be done further, so that this can finally be merged and not rollbacked?

**Comment by twaik** [comment]
> Why is this issue closed, if the commit regarding this issue ([2104ea6](https://github.com/termux/termux-x11/commit/2104ea690d3f043d381f2ca9c6e6cfef723b4ca5)) was reverted back via this commit ([76309c1](https://github.com/termux/termux-x11/commit/76309c1c7b7a515a942a650d2b8e8e40fd06c865))??

Probably it was removed by mistake. Now it is back.

**Comment by twaik** [comment]
> Are there any specific guidelines as to what is acceptable & what is not, regarding the code

First of all changes should not break behaviour of input handling on different devices. If there will be anything else I'll tell you. 

I do not think we should create new issue for this.

**Comment by hansm629** [comment]
@twaik 
Based on build version #590

When using the touchpad of the Galaxy Tab S9 Ultra keyboard cover on the X11 server, there is still a symptom of the touchpad moving uncontrollably quickly. T^T

https://github.com/termux/termux-x11/assets/101167173/2912f0cb-f25e-41c2-8882-82ba563e1bca

**Comment by twaik** [comment]
Ok, there is new preference called "captured pointer speed factor" in [build 591](https://github.com/termux/termux-x11/actions/runs/8424425295).

**Comment by hansm629** [comment]
@twaik 
Sorry for the late feedback.

As a result of testing with build #593, we confirmed that the touchpad axis was still reversed and moved very quickly.

**Comment by twaik** [comment]
> axis was still reversed and moved very quickly.

Yeah, and that is a reason I made two new preferences for that. Did you test them?

**Comment by hansm629** [comment]
@twaik 
Are you talking about the pointer speed control feature below?
Once set to 0, the pointer speed becomes normal.

However, the speed of the mouse connected via Bluetooth seems to be very slow.
![Screenshot_20240326_132456_TermuxX11](https://github.com/termux/termux-x11/assets/101167173/2cc5324f-d740-4f1e-9e57-a473bbe3298d)

**Comment by hansm629** [comment]
@twaik 
And strangely enough, if you apply the function to invert the pointer axis, the pointer axis of the mouse connected via Bluetooth will change, but

The touchpad on the keyboard cover did not change.

**Comment by twaik** [comment]
I was talking about both "transform captured pointer mouse" and "captured pointer speed factor". These preferences are not device specific, you have to change them according to the device you use. If you have very slow pointer of bt mouse you have to set speed factor to 200.

**Comment by twaik** [comment]
I'll fix that.

**Comment by twaik** [comment]
> @twaik Are you talking about the pointer speed control feature below? Once set to 0, the pointer speed becomes normal.
> 
> However, the speed of the mouse connected via Bluetooth seems to be very slow. ![Screenshot_20240326_132456_TermuxX11](https://private-user-images.githubusercontent.com/101167173/316726886-2cc5324f-d740-4f1e-9e57-a473bbe3298d.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE0Mjk3NTcsIm5iZiI6MTcxMTQyOTQ1NywicGF0aCI6Ii8xMDExNjcxNzMvMzE2NzI2ODg2LTJjYzUzMjRmLWQ3NDAtNGYxZS05ZTU3LWE0NzNiYmUzMjk4ZC5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMyNlQwNTA0MTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iZTVhMGFjNWQzMDViYTJiOTVkNDJhNzMzMzc4ZTA1YTMyYzQwNGY0MWJjZTg1MTcwNTllN2ZhMWI0YzQzMzk4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.ll3R2dllghFhA5e2vrRSzbaJbpaG7IZic2w-VsJ1Prg)

Minimum is 30%, not 0%.
